### PR TITLE
Fix #9947: make the ABI identical between debug and non-debug builds

### LINKED ISSUE #9947: ABI may depend on NDEBUG
In changeset ab4585a695, the `InternalMetadata` destructor [changed](https://github.com/protocolbuffers/protobuf/commit/ab4585a6956675ce14a1cba5d321fde980bbf12b#diff-5463672866624dbb2eb60cc9cd9282520f6fd68e78414e2ce17cdf1f1520fbbaR73) to the following:
```c++
#if defined(NDEBUG) || defined(_MSC_VER)
  ~InternalMetadata() {
    if (HasMessageOwnedArenaTag()) {
      delete reinterpret_cast<Arena*>(ptr_ - kMessageOwnedArenaTagMask);
    }
  }
#else
  ~InternalMetadata();
#endif
```

This causes ABI problems between debug and non-debug in one specific case: when the `-fvisibility-inlines-hidden` compiler option is being used. According to the [GCC wiki](https://gcc.gnu.org/wiki/Visibility), that option has the following effect:
> This causes all inlined class member functions to have hidden visibility, causing significant export symbol table size & binary size reductions

However, when `libprotobuf.so` was compiled in non-debug mode with `-fvisibility-inlines-hidden` enabled, any debug build of a third-party application will fail looking up `~InternalMetadata()`

**What version of protobuf and what language are you using?**
Version: v3.20.1
Language: C++

**What runtime / compiler are you using (e.g., python version or gcc version)**
gcc 10.3.0

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
cc @fowles @coryan

File: src/google/protobuf/metadata_lite.h [hunk 1]
```diff
@@ -77,15 +77,19 @@ class PROTOBUF_EXPORT InternalMetadata {
     GOOGLE_DCHECK(!is_message_owned || arena != nullptr);
   }
 
-#if defined(NDEBUG) || defined(_MSC_VER)
+  // To keep the ABI identical between debug and non-debug builds,
```
[the remark this answers is not in the document]
Comment (pitrou) [reply src/google/protobuf/metadata_lite.h:80]: Yes, it was already stated on the linked issue as a reason to not bother with the ABI problem.
However, as you also noted, all Unix compiler pretty much have to deal nicely with this as the practice of building in debug mode against non-debug dependencies is almost universal. So the compiler/linker will end up selecting one implementation over another; which one exactly doesn't really matter in practice (the only difference being the added debug checks).

**Comment by fowles** [comment]
We have a 21.x release coming later this week (god willing and the creek don't rise).  I will make sure this gets into it.

**Comment by fowles** [comment]
https://github.com/protocolbuffers/protobuf/pull/10272 should have fixed the distcheck errors so you will need to rebase on top of that

**Comment by pitrou** [comment]
Ok, I rebased.

**Comment by fowles** [comment]
No problem.  I do want to be clear that this is never a configuration we will officially support.  We will continue to review PRs on a best effort basis for it, but you are in the danger zone.  I am sorry that a bunch of my earlier communication on this made it seem like a very hard "no" and not more of "you are in the land of dragons and we will only go so far with you".

**Comment by knedlsepp** [comment]
> No problem. I do want to be clear that this is never a configuration we will officially support. We will continue to review PRs on a best effort basis for it, but you are in the danger zone. I am sorry that a bunch of my earlier communication on this made it seem like a very hard "no" and not more of "you are in the land of dragons and we will only go so far with you".

It sounds like this requirement could be made explicit by ensuring that `-DNDEBUG` is part of the `pkg-config --cflags protobuf` output in case protobuf was built with that. This would match quite well with what is already in the README:

[> To compile a package that uses Protocol Buffers, you need to pass various flags to your compiler and linker.](https://github.com/protocolbuffers/protobuf/blob/3b456bfcd814941baec4fd55400438d0ab6e909a/src/README.md?plain=1#L79-L81)

Apparently `-DNDEBUG` should be part of the flags mentioned there.

**Comment by adamnovak** [comment]
> Apparently -DNDEBUG should be part of the flags mentioned there.

Wouldn't that make it quite difficult to use `assert()` in user code?

**Comment by fowles** [comment]
@adamnovak ideally, when compiling with asserts enabled you would also compile protobuf with asserts enabled.  (We actually have a bunch of extra asserts that will help users catch bugs in their handling of protobuffers)

**Comment by knedlsepp** [comment]
> Wouldn't that make it quite difficult to use `assert()` in user code?

Yes, and for that matter I'm very grateful that @fowles opted for still best-effort supporting this.
However if there were no negative side-effect to adding this flag for downstream users and it being a *requirement* for the protobuf build that downstream users must use the same flags that were used at build-time then `pkg-config` would be the way to go.

*Ideally* protobuf would not choose a macro that clashes with downstream usage though.
Maybe something like `-DPROTOBUF_OPTIMIZED_BUILD`, which would then be free to ignore ABI-stability between debug and release builds.

**Comment by fowles** [comment]
I am not opposed to having a `PROTOBUF_NDEBUG` that is independent of `NDEBUG`.  Unfortunately, that work would have to be done by a googler (not necessarily someone on the protobuf team).  Protobuf's code is a bit of an iceberg where there is a large volume of google internal code.  PRs that fixed this externally would leave our code base in a worse franken state.

**Comment by adamnovak** [comment]
I think Protobuf could still keep using the NDEBUG macro name internally if there were a way to make sure it had the right state in the Protobuf headers, independent of the state it needs to have in the application code that includes them.

One way to do this would be to frame the places where the ABI depends on `NDEBUG` with macros that make sure `NDEBUG` is defined or not defined so as to match what the library we will need to link against, while keeping changes from leaking out of the Protobuf headers into client code.

Something like:

Leading header (include after all other Protobuf headers in Protobuf headers with variable ABIs):
```
#ifndef PROTOBUF_ABI_COMPAT_ACTIVE
    #define PROTOBUF_ABI_COMPAT_ACTIVE
    #if defined(PROTOBUF_LIBRARY_IS_DEBUG) && defined(NDEBUG)
        // App is release, but needs to link against a debug Protobuf built without NDEBUG    
        #define PROTOBUF_RELEASE_APP_ON_DEBUG_LIBRARY
        #undef NDEBUG
    #elif defined(PROTOBUF_LIBRARY_IS_RELEASE) && !defined(NDEBUG)
        // App is debug but needs to link against a release protobuf built with NDEBUG
        #define PROTOBUF_DEBUG_APP_ON_RELEASE_LIBRARY
        #define NDEBUG
    #endif
#endif
```

Trailing header (include at the end of all Protobuf headers that include the leading header):
```
#ifdef PROTOBUF_ABI_COMPAT_ACTIVE
    #undef PROTOBUF_ABI_COMPAT_ACTIVE
    #if defined(PROTOBUF_RELEASE_APP_ON_DEBUG_LIBRARY)
        #undef PROTOBUF_RELEASE_APP_ON_DEBUG_LIBRARY
        #define NDEBUG
    #elif defined(PROTOBUF_DEBUG_APP_ON_RELEASE_LIBRARY)
        #undef PROTOBUF_DEBUG_APP_ON_RELEASE_LIBRARY
        #undef NDEBUG
    #endif
#endif
```

Then in use it would look like:

```

// Other includes

#include <google/protobuf/abi_compat_on.h>

class Whatsit {
#ifdef NDEBUG
    // One ABI for release
    void some_method();
#else
    // Another completely different ABI for debug
    int some_method();
#endif
};

#include <google/protobuf/abi_compat_off.h>

```

If you set it up like that, `NDEBUG` still controls the ABI if no additional macros are defined, or in Google-internal headers that haven't yet been changed. But an application build system can decide if it is dealing with a release or debug Protobuf library to link against, and send `-DPROTOBUF_LIBRARY_IS_DEBUG` or `-DPROTOBUF_LIBRARY_IS_RELEASE` as appropriate, and then the Protobuf headers will make sure to adopt the right `NDEBUG` setting to match the library, without letting it leak out when they themselves are included in user code.

Then if `pkg-config` wants to communicate information about the CPPFLAGS to use with the Protobuf library, it could then do it via Protobuf-namespaced macro definitions and not grab hold of the global `NDEBUG`.

Whatever the final answer is, it probably would want to work well with [CMake's `FindProtobuf`](https://cmake.org/cmake/help/latest/module/FindProtobuf.html) in addition to just `pkg-config`, which might involve making contributions to CMake. It [already knows how to find and manage debug and release versions of the library](https://github.com/Kitware/CMake/blob/96564c5b2d84568459c2b594a7db3baccb524260/Modules/FindProtobuf.cmake#L529-L565), but it also has just a default version that could be either, that it will let you try to link against.

**Pull request merged**